File: src/main/java/cc/mirukuneko/realtrainmodrenewed/client/TrainHudOverlay.java
package cc.mirukuneko.realtrainmodrenewed.client;

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed;
import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity;
import cc.mirukuneko.realtrainmodrenewed.entity.TrainSeatEntity;
import cc.mirukuneko.realtrainmodrenewed.vehicle.VehicleDefinition;
import cc.mirukuneko.realtrainmodrenewed.vehicle.VehicleRegistry;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.renderer.RenderPipelines;
import net.minecraft.resources.Identifier;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.client.event.RenderGuiEvent;

@EventBusSubscriber(modid = RealTrainModRenewed.MODID, value = Dist.CLIENT)
public final class TrainHudOverlay {
    private static final Identifier CAB_TEXTURE =
        Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "textures/gui/rtm_cab.png");
    private static final int TEX_SIZE = 512;
    private static final int CAB_W = 416;
    private static final int CAB_H = 48;
    private static boolean cabHidden;

    private TrainHudOverlay() {
    }

    public static void toggleCabHidden() {
        cabHidden = !cabHidden;
    }

    @SubscribeEvent
    public static void onRenderGui(RenderGuiEvent.Post event) {
        Minecraft mc = Minecraft.getInstance();
        if (mc.player == null || mc.screen != null || mc.options.hideGui) {
            return;
        }
        TrainEntity train = getControlledTrain(mc);
        if (train == null || !train.isDriverPassenger(mc.player)) {
            return;
        }

        GuiGraphicsExtractor g = event.getGuiGraphics();
        Font font = mc.font;
        int screenW = mc.getWindow().getGuiScaledWidth();
        int screenH = mc.getWindow().getGuiScaledHeight();
        VehicleDefinition def = VehicleRegistry.getById(train.getVehicleId());
        boolean showCabOverlay = def == null || !def.isNotDisplayCab();

        if (!cabHidden && showCabOverlay) {
            renderDefaultRtmCab(g, font, train, def, screenW, screenH);
        }
    }

    private static void renderDefaultRtmCab(GuiGraphicsExtractor graphics, Font font, TrainEntity train,
                                            VehicleDefinition def, int screenW, int screenH) {
        float scale = Math.min(1.0F, screenW / (float) CAB_W);
        int x = Math.round((screenW - CAB_W * scale) * 0.5F);
        int y = Math.round(screenH - CAB_H * scale);
        graphics.blit(RenderPipelines.GUI_TEXTURED, CAB_TEXTURE, x, y, 0.0F, 0.0F, Math.round(CAB_W * scale), Math.round(CAB_H * scale), TEX_SIZE, TEX_SIZE);
        drawMeter(graphics, 32, 19, 32, 32, 48, 240.0F * getBrakeRatio(train));
        drawMeter(graphics, 32, 19, 32, 0, 48, 240.0F * getBrakeCommandRatio(train));
        drawMeter(graphics, 72, 19, 32, 64, 48, getSpeedNeedleRotation(train, def));
        drawLever(graphics, 104, 29, train);
        drawWatch(graphics, train);
        graphics.text(font, Integer.toString(getSpeedKmh(train)), x + 70, y + 37, 0x00FF00, false);
        // ブレーキ段数表示 (B1-B8)。本家同様ノッチ番号をそのまま出す。
        graphics.text(font, Integer.toString(Math.max(0, -train.getNotch())), x + 30, y + 37, 0x00FF00, false);
        graphics.text(font, Integer.toString(getWorldTime()), x + 338, y + 8, 0x00FF00, false);
        graphics.text(font, getClockText(), x + 338, y + 18, 0x00FF00, false);
    }

    private static TrainEntity getControlledTrain(Minecraft mc) {
        if (mc.player == null) {
            return null;
        }
        if (mc.player.getVehicle() instanceof TrainEntity train) {
            return train;
        }
        if (mc.player.getVehicle() instanceof TrainSeatEntity seat) {
            return seat.getTrain();
        }
        return null;
    }

    private static void drawLever(GuiGraphicsExtractor graphics, int x, int y, TrainEntity train) {
        int notch = train.getNotch();
        // rtm_cab.png のマスコン目盛りは中立(y28)から 3px 等間隔で並ぶ(実測):
        //   EB(-8)=y4(赤), B7=y7 ... B1=y25, N=y28, P1=y31 ... P5=y43。
        // よって針位置は notch に対して線形 y = 28 + 3*notch。本家RTMと同じ等間隔の針送りになる。
        float offset = 3.0F * notch;
        graphics.blit(RenderPipelines.GUI_TEXTURED, CAB_TEXTURE, x - 4, Math.round(y + offset - 2), 0.0F, 80.0F, 8, 3, TEX_SIZE, TEX_SIZE);
    }

    private static void drawWatch(GuiGraphicsExtractor graphics, TrainEntity train) {
        int startX = 320;
        int startY = 32;
        int t0 = getWorldTime(train);
        int hour12 = (t0 / 1000 + 6) % 12;
        drawMeter(graphics, startX, startY, 32, 96, 48, 360.0F * hour12 / 12.0F + 135.0F);
        int minute = (int) ((t0 % 1000) * 0.06F);
        drawMeter(graphics, startX, startY, 32, 128, 48, 360.0F * minute / 60.0F + 135.0F);
    }

    private static void drawWatch(GuiGraphicsExtractor graphics) {
        drawMeter(graphics, 320, 32, 32, 96, 48, 135.0F);
        drawMeter(graphics, 320, 32, 32, 128, 48, 135.0F);
    }

    private static void drawMeter(GuiGraphicsExtractor graphics, int x, int y, int size, int u, int v, float rotation) {
        int offset = -(size / 2);
        graphics.blit(RenderPipelines.GUI_TEXTURED, CAB_TEXTURE, x + offset, y + offset, u, v, size, size, TEX_SIZE, TEX_SIZE);
    }

    private static int getSpeedKmh(TrainEntity train) {
        return Math.round(Math.abs(train.getSpeed()) * 72.0F);
    }

    private static float getSpeedNeedleRotation(TrainEntity train, VehicleDefinition def) {
        float max = 120.0F;
        if (def != null && !def.getNotchMaxSpeeds().isEmpty()) {
            for (Float speed : def.getNotchMaxSpeeds()) {
                if (speed != null) {
                    max = Math.max(max, speed);
                }
            }
        }
        return Math.min(270.0F, 270.0F * getSpeedKmh(train) / Math.max(1.0F, max));
    }

    private static float getBrakeRatio(TrainEntity train) {
        // 実際の最大ブレーキ段数で割る(段数とメーターのズレを防ぐ)。
        return Math.min(1.0F, Math.max(0.0F, -train.getNotch()) / (float) Math.max(1, train.getMaxBrakeNotch()));
    }

    private static float getBrakeCommandRatio(TrainEntity train) {
        return Math.min(1.0F, Math.max(0.0F, -train.getNotch()) / (float) Math.max(1, train.getMaxBrakeNotch()));
    }

    private static int getWorldTime() {
        Minecraft mc = Minecraft.getInstance();
        return mc.level == null ? 0 : (int) (mc.level.getLevelData().getGameTime() % 24000L);
    }

    private static int getWorldTime(TrainEntity train) {
        return train.level() == null ? getWorldTime() : (int) (train.level().getLevelData().getGameTime() % 24000L);
    }

    private static String getClockText() {
        int t0 = getWorldTime();
        int hour = (t0 / 1000 + 6) % 24;
        int minute = (int) ((t0 % 1000) * 0.06F);
        return hour + ":" + minute;
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/client/TrainHudOverlay.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.client

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity
import cc.mirukuneko.realtrainmodrenewed.entity.TrainSeatEntity
import cc.mirukuneko.realtrainmodrenewed.vehicle.VehicleDefinition
import cc.mirukuneko.realtrainmodrenewed.vehicle.VehicleRegistry
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.renderer.RenderPipelines
import net.minecraft.resources.Identifier
import net.neoforged.api.distmarker.Dist
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.fml.common.EventBusSubscriber
import net.neoforged.neoforge.client.event.RenderGuiEvent
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sin

@EventBusSubscriber(modid = RealTrainModRenewed.MODID, value = [Dist.CLIENT])
object TrainHudOverlay {
    private val CAB_TEXTURE: Identifier =
        Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "textures/gui/rtm_cab.png")
    private const val TEX_SIZE = 512
    private const val CAB_W = 416
    private const val CAB_H = 48
    private const val HUD_GREEN = 0xFF40FF40.toInt()
    private const val HUD_RED = 0xFFE02020.toInt()
    private const val HUD_DARK = 0xFF303030.toInt()
    private var cabHidden = false

    @JvmStatic
    fun toggleCabHidden() {
        cabHidden = !cabHidden
    }

    @JvmStatic
    @SubscribeEvent
    fun onRenderGui(event: RenderGuiEvent.Post) {
        val mc = Minecraft.getInstance()
        if (mc.player == null || mc.screen != null || mc.options.hideGui) {
            return
        }
        val train = getControlledTrain(mc)
        if (train == null || !train.isDriverPassenger(mc.player)) {
            return
        }

        val graphics = event.guiGraphics
        val font = mc.font
        val screenW = mc.window.guiScaledWidth
        val screenH = mc.window.guiScaledHeight
        val definition = VehicleRegistry.getById(train.vehicleId)
        val showCabOverlay = definition == null || !definition.isNotDisplayCab()

        if (!cabHidden && showCabOverlay) {
            renderDefaultRtmCab(graphics, font, train, definition, screenW, screenH)
        }
    }

    private fun renderDefaultRtmCab(
        graphics: GuiGraphicsExtractor,
        font: Font,
        train: TrainEntity,
        definition: VehicleDefinition?,
        screenW: Int,
        screenH: Int,
    ) {
        val scale = min(1.0f, screenW / CAB_W.toFloat())
        val x = ((screenW - CAB_W * scale) * 0.5f).roundToInt()
        val y = (screenH - CAB_H * scale).roundToInt()
        graphics.blit(
            RenderPipelines.GUI_TEXTURED,
            CAB_TEXTURE,
            x,
            y,
            0.0f,
            0.0f,
            (CAB_W * scale).roundToInt(),
            (CAB_H * scale).roundToInt(),
            TEX_SIZE,
            TEX_SIZE,
        )
        drawGaugeNeedle(graphics, x, y, scale, 32, 19, 13.0f, getGaugeAngle(240.0f * getBrakeRatio(train)), HUD_RED)
        drawGaugeNeedle(graphics, x, y, scale, 32, 19, 9.0f, getGaugeAngle(240.0f * getBrakeCommandRatio(train)), HUD_DARK)
        drawGaugeNeedle(graphics, x, y, scale, 72, 19, 13.0f, getGaugeAngle(getSpeedNeedleRotation(train, definition)), HUD_RED)
        drawLever(graphics, x, y, scale, train)
        drawWatch(graphics, x, y, scale, train)
        drawCenteredText(graphics, font, getSpeedKmh(train).toString(), x, y, scale, 72, 37)
        drawCenteredText(graphics, font, max(0, -train.notch).toString(), x, y, scale, 32, 37)
        graphics.text(font, getWorldTime().toString(), scaledX(x, scale, 338), scaledY(y, scale, 8), HUD_GREEN, true)
        graphics.text(font, getClockText(), scaledX(x, scale, 338), scaledY(y, scale, 18), HUD_GREEN, true)
    }

    private fun getControlledTrain(mc: Minecraft): TrainEntity? {
        val player = mc.player ?: return null
        val vehicle = player.vehicle
        if (vehicle is TrainEntity) {
            return vehicle
        }
        if (vehicle is TrainSeatEntity) {
            return vehicle.getTrain()
        }
        return null
    }

    private fun drawLever(graphics: GuiGraphicsExtractor, x: Int, y: Int, scale: Float, train: TrainEntity) {
        val offset = 3.0f * train.notch
        val pose = graphics.pose()
        pose.pushMatrix()
        pose.translate(x.toFloat(), y.toFloat())
        pose.scale(scale, scale)
        graphics.blit(RenderPipelines.GUI_TEXTURED, CAB_TEXTURE, 100, (27 + offset).roundToInt(), 0.0f, 80.0f, 8, 3, TEX_SIZE, TEX_SIZE)
        pose.popMatrix()
    }

    private fun drawWatch(graphics: GuiGraphicsExtractor, x: Int, y: Int, scale: Float, train: TrainEntity) {
        val startX = 320
        val startY = 32
        val t0 = getWorldTime(train)
        val hour12 = (t0 / 1000 + 6) % 12
        drawMeter(graphics, x, y, scale, startX, startY, 32, 96, 48, 360.0f * hour12 / 12.0f + 135.0f)
        val minute = ((t0 % 1000) * 0.06f).toInt()
        drawMeter(graphics, x, y, scale, startX, startY, 32, 128, 48, 360.0f * minute / 60.0f + 135.0f)
    }

    private fun drawMeter(
        graphics: GuiGraphicsExtractor,
        x: Int,
        y: Int,
        scale: Float,
        localX: Int,
        localY: Int,
        size: Int,
        u: Int,
        v: Int,
        rotation: Float,
    ) {
        val pose = graphics.pose()
        pose.pushMatrix()
        pose.translate(x + localX * scale, y + localY * scale)
        pose.rotate(Math.toRadians(rotation.toDouble()).toFloat())
        pose.scale(scale, scale)
        val offset = -(size / 2)
        graphics.blit(RenderPipelines.GUI_TEXTURED, CAB_TEXTURE, offset, offset, u.toFloat(), v.toFloat(), size, size, TEX_SIZE, TEX_SIZE)
        pose.popMatrix()
    }

    private fun drawGaugeNeedle(
        graphics: GuiGraphicsExtractor,
        x: Int,
        y: Int,
        scale: Float,
        localX: Int,
        localY: Int,
        length: Float,
        angleDegrees: Float,
        color: Int,
    ) {
        val cx = scaledX(x, scale, localX)
        val cy = scaledY(y, scale, localY)
        val radians = Math.toRadians(angleDegrees.toDouble())
        val dx = cos(radians)
        val dy = sin(radians)
        val steps = max(1, (length * scale).roundToInt())
        for (i in 0..steps) {
            val t = i / steps.toDouble()
            val px = (cx + dx * length * scale * t).roundToInt()
            val py = (cy + dy * length * scale * t).roundToInt()
            graphics.fill(px, py, px + 1, py + 1, color)
        }
        graphics.fill(cx - 1, cy - 1, cx + 2, cy + 2, 0xFFE8E8E8.toInt())
    }

    private fun getGaugeAngle(rotation: Float): Float = 150.0f + rotation

    private fun scaledX(x: Int, scale: Float, localX: Int): Int =
        (x + localX * scale).roundToInt()

    private fun scaledY(y: Int, scale: Float, localY: Int): Int =
        (y + localY * scale).roundToInt()

    private fun drawCenteredText(
        graphics: GuiGraphicsExtractor,
        font: Font,
        text: String,
        x: Int,
        y: Int,
        scale: Float,
        localX: Int,
        localY: Int,
    ) {
        val textX = scaledX(x, scale, localX) - font.width(text) / 2
        val textY = scaledY(y, scale, localY)
        graphics.fill(textX - 2, textY - 1, textX + font.width(text) + 2, textY + font.lineHeight, 0xCC000000.toInt())
        graphics.text(font, text, textX, textY, HUD_GREEN, true)
    }

    private fun getSpeedKmh(train: TrainEntity): Int =
        (abs(train.speed) * 72.0f).roundToInt()

    private fun getSpeedNeedleRotation(train: TrainEntity, definition: VehicleDefinition?): Float {
        var maxSpeed = 120.0f
        if (definition != null && definition.notchMaxSpeeds.isNotEmpty()) {
            for (speed in definition.notchMaxSpeeds) {
                maxSpeed = max(maxSpeed, speed)
            }
        }
        return min(270.0f, 270.0f * getSpeedKmh(train) / max(1.0f, maxSpeed))
    }

    private fun getBrakeRatio(train: TrainEntity): Float =
        min(1.0f, max(0.0f, train.brakeCylinderPressure / 480.0f))

    private fun getBrakeCommandRatio(train: TrainEntity): Float =
        min(1.0f, max(0.0f, -train.notch.toFloat()) / max(1, train.maxBrakeNotch).toFloat())

    private fun getWorldTime(): Int {
        val mc = Minecraft.getInstance()
        return (mc.level?.levelData?.gameTime?.rem(24000L) ?: 0L).toInt()
    }

    private fun getWorldTime(train: TrainEntity): Int =
        (train.level().levelData.gameTime % 24000L).toInt()

    private fun getClockText(): String {
        val t0 = getWorldTime()
        val hour = (t0 / 1000 + 6) % 24
        val minute = ((t0 % 1000) * 0.06f).toInt()
        return "$hour:$minute"
    }
}
